tex-fmt: add module

WGUNDERWOOD opened the pull request:
tex-fmt is a LaTeX source code formatter written in Rust, and uses a user configuration file in the .toml format.

### Description

<!--

Please provide a brief description of your change.

-->

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->

WGUNDERWOOD commented:
This is a draft PR which is waiting for https://github.com/NixOS/nixpkgs/pull/363212 to be merged.

mathjiajia commented:
any update?

mirkolenz reviewed (commented):
Just came across this this PR after creating a tex-fmt module for myself 😄 First of all, thank you very much for this great tool, it is so much better than latexindent!

Regarding the proposed module, I have some suggestions that could improve the interface:

- Instead of using the name `options` for custom settings, I would advocate for the name `settings` instead. The former is used to define nix module options, meaning the name could lead to confusion for some users. Using the name `settings` has also been adopted by nixpkgs: <https://github.com/NixOS/rfcs/pull/42>.
- According to the documentation, tex-fmt does not use XDG directories on macOS but instead uses the application support dir. If so, the current module would only work on Linux.
- Since you're using toml as a config language, you can use the built-in nixpkgs formats that have corresponding option types.
- The use of `with lib` in the top-level has been discouraged.
- It would be good to have some tests making sure that the config file is written correctly.

If you're interested, here is my own module. Feel free to copy snippets of it as needed (you can also add me as a co-maintainer if you'd like to). Please note that it is formatted using the new nixfmt rfc style that is not yet used in home manager.

```nix
{
  pkgs,
  config,
  lib,
  ...
}:

let

  inherit (lib)
    mkEnableOption
    mkPackageOption
    mkOption
    literalExpression
    ;

  configDir = if pkgs.stdenv.isDarwin then "Library/Application Support" else config.xdg.configHome;

  tomlFormat = pkgs.formats.toml { };
  cfg = config.programs.tex-fmt;

in
{
  meta.maintainers = with lib.maintainers; [ mirkolenz ];

  options.programs.tex-fmt = {
    enable = mkEnableOption "tex-fmt";

    package = mkPackageOption pkgs "tex-fmt" { };

    settings = mkOption {
      type = tomlFormat.type;
      default = { };
      example = literalExpression ''
        {
          wrap = false;
          tabsize = 2;
          tabchar = "space";
          lists = [];
        }
      '';
      description = ''
        Configuration written to
        {file}`$XDG_CONFIG_HOME/tex-fmt/tex-fmt.toml` on Linux or
        {file}`$HOME/Library/Application Support/tex-fmt/tex-fmt.toml` on Darwin.
        See <https://github.com/WGUNDERWOOD/tex-fmt> and
        <https://github.com/WGUNDERWOOD/tex-fmt/blob/master/tex-fmt.toml>
        for more information.
      '';
    };
  };

  config = lib.mkIf cfg.enable {
    home.packages = [ cfg.package ];

    home.file."${configDir}/tex-fmt/tex-fmt.toml" = lib.mkIf (cfg.settings != { }) {
      source = tomlFormat.generate "tex-fmt-config" cfg.settings;
    };
  };
}
```

WGUNDERWOOD commented:
Thanks for these pointers! I will change `options` to `settings`. Note that the config directory is set within tex-fmt using [`dirs::config_dir`](https://docs.rs/dirs/latest/dirs/fn.config_dir.html); perhaps this is a known issue if it disagrees with Nix on MacOS? I am also unsure how much example config to provide: I want to show the user how to pass options in `tex-fmt.toml`, but also want to avoid adding too much to simplify maintenance in the future, should any options change.

mirkolenz commented:
> perhaps this is a known issue if it disagrees with Nix on MacOS

I think it's not so much related to nix on macOS but rather a difference between macOS and Linux. There are a few other applications (e.g., poetry) that behave the same way. I know that for instance uv supports both XDG config directories and the Application Support directory on macOS, but I don't know if they have custom logic in place. In the end, I don't think it's too much of an issue because this workaround I added here is also used in some other home manager modules.

>  I am also unsure how much example config to provide

I think it's sufficient to have one or two entries in here together with the links to the documentation. I just added my own config as an example here 😄

WGUNDERWOOD commented:
Great, thanks for this. Could you explain why you have semicolons at the end of each line in the example config? As far as I know, this is not needed in toml.

WGUNDERWOOD commented:
I see, thanks I just thought they might be removable since it's inside a `literalExpression`, but not if it's being parsed as Nix. I've basically just taken your example and added myself as a maintainer and changed the default settings slightly.

khaneliman commented on modules/programs/tex-fmt.nix:
```diff
@@ -0,0 +1,53 @@
+
+  inherit (lib) mkEnableOption mkPackageOption mkOption literalExpression;
+
+  configDir = if pkgs.stdenv.isDarwin then
```
Have you tested darwin with xdg enabled? Does it still write to `Library/Application Support` ?

khaneliman replied:
Sorry, was reading above and it seems like you were talking about it that they might just ignore XDG on macos. If that's not the case, it can be fixed in a follow up.
